// @redocly/respect-core/lib/modules/faker.js
import { faker } from '@faker-js/faker';
export function createFaker() {
    const fakeString = {
        email: ({ provider, domain = 'com' } = {}) => faker.internet.email(undefined, undefined, `${provider}.${domain}`),
        userName: () => faker.internet.userName(),
        firstName: () => faker.name.firstName(),
        lastName: () => faker.name.lastName(),
        fullName: () => faker.name.fullName(),
        uuid: () => faker.datatype.uuid(),
        string: ({ length } = {}) => faker.datatype.string(length),
    };
    const fakeDate = {
        past: () => faker.date.past(),
        future: () => faker.date.future(),
    };
    const fakeAddress = {
        city: () => faker.address.city(),
        country: () => faker.address.country(),
        zipCode: () => faker.address.zipCode(),
        street: () => faker.address.street(),
    };
    const fakeNumber = {
        integer: ({ min, max } = {}) => faker.datatype.number({ min, max, precision: 1 }),
        float: (options) => faker.datatype.float(options),
    };
    return {
        address: fakeAddress,
        date: fakeDate,
        number: fakeNumber,
        string: fakeString,
    };
}
