// jsonpath-rfc9535/dist/commonjs/core/exec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = exec;
const index_ts_1 = __importDefault(require("../parser/index.js"));
const stack_ts_1 = require("../utils/stack.js");
const count_ts_1 = __importDefault(require("./functions/count.js"));
const length_ts_1 = __importDefault(require("./functions/length.js"));
const match_ts_1 = __importDefault(require("./functions/match.js"));
const search_ts_1 = __importDefault(require("./functions/search.js"));
const value_ts_1 = __importDefault(require("./functions/value.js"));
const query_ts_1 = __importDefault(require("./visitors/query.js"));
function exec(input, expression, opts, cb) {
    const ctx = {
        cache: new Map(),
        capturePaths: opts.capturePaths,
        functions: {
            count: count_ts_1.default,
            search: search_ts_1.default,
            value: value_ts_1.default,
            match: match_ts_1.default,
            length: length_ts_1.default,
        },
        regexp: "i-regexp",
        stack: new stack_ts_1.Stack(),
    };
    (0, query_ts_1.default)(ctx, input, input, (0, index_ts_1.default)(expression), cb);
}
