// jsonpath-rfc9535/dist/commonjs/core/visitors/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = visitQuery;
const stack_ts_1 = require("../../utils/stack.js");
const path_ts_1 = require("../path.js");
const segment_ts_1 = require("./segment.js");
function visitQuery(ctx, root, input, node, cb) {
    const stack = new stack_ts_1.Stack();
    stack.push({
        root,
        path: (0, path_ts_1.getInitialPath)(ctx.capturePaths),
        value: input,
        index: 0,
    });
    const context = {
        cache: ctx.cache,
        capturePaths: false,
        functions: ctx.functions,
        regexp: ctx.regexp,
        stack,
    };
    while (stack.size > 0) {
        // biome-ignore lint/style/noNonNullAssertion: size is guaranteed to be > 0
        const item = stack.pop();
        if (item.index === node.segments.length) {
            cb(item.value, item.path);
            continue;
        }
        const { path, value, index } = item;
        const segment = node.segments[index];
        (0, segment_ts_1.visitSegment)(context, item, segment.node);
        if (segment.type === "DescendantSegment") {
            if (Array.isArray(value)) {
                for (let i = 0; i < value.length; i++) {
                    stack.push({
                        root,
                        path: (0, path_ts_1.joinPathWithKey)(path, i),
                        value: value[i],
                        index,
                    });
                }
            }
            else if (value !== null && typeof value === "object") {
                for (const key of Object.keys(value)) {
                    stack.push({
                        root,
                        path: (0, path_ts_1.joinPathWithKey)(path, key),
                        value: value[key],
                        index,
                    });
                }
            }
        }
    }
}
