// jsonpath-rfc9535/dist/commonjs/core/functions/length.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const guards_ts_1 = require("../../utils/guards.js");
const results_ts_1 = require("../results.js");
// Any Unicode code point except high-surrogate and low-surrogate code points.
function countUnicodeScalarValues(str) {
    let count = 0;
    for (let i = 0; i < str.length; i++) {
        const code = str.charCodeAt(i);
        if ((code >= 0 && code <= 0xd7ff) || (code >= 0xe000 && code <= 0x10ffff)) {
            count++;
        }
    }
    return count;
}
exports.default = {
    declaration: function length(ctx, value) {
        if (typeof value === "string") {
            return countUnicodeScalarValues(value);
        }
        if (Array.isArray(value)) {
            return value.length;
        }
        if ((0, guards_ts_1.isPlainObject)(value)) {
            return Object.keys(value).length;
        }
        return results_ts_1.Nothing;
    },
    definition: {
        parameters: ["ValueType"],
        returnType: "ValueType",
    },
};
